x-pack/plugins/visualization_lens/public/types.ts

```diff
@@ -0,0 +1,176 @@
+  registerVisualization: (visualization: Visualization<unknown>) => void;
+}
+
+export interface EditorFrameState {
+  visualizations: { [id: string]: object };
+  datasources: { [id: string]: object };
+
+  activeDatasourceId: string;
+  activeVisualizationId: string;
+}
+
+// Hints the default nesting to the data source. 0 is the highest priority
+export type DimensionPriority = 0 | 1 | 2;
+
+// For switching between visualizations and correctly matching columns
+export type DimensionRole =
+  | 'splitChart'
+  | 'series'
+  | 'primary'
+  | 'secondary'
+  | 'color'
+  | 'size'
+  | string; // Some visualizations will use custom names that have other meaning
+
+export interface TableMetaInfo {
+  columnId: string;
+  operation: Operation;
+}
+
+export interface DatasourceSuggestion<T> {
+  state: T;
+  tableMetas: TableMetaInfo[];
+}
+
+/**
+ * Interface for the datasource registry
+ */
+export interface Datasource<T> {
+  // For initializing from saved object
+  // initialize: (state?: T) => Promise<T>;
+
+  renderDataPanel: (props: DatasourceDataPanelProps) => void;
+
+  toExpression: (state: T) => string;
+
+  getDatasourceSuggestionsForField: (state: T) => Array<DatasourceSuggestion<T>>;
+  getDatasourceSuggestionsFromCurrentState: (state: T) => Array<DatasourceSuggestion<T>>;
+
+  getPublicAPI: (state: T, setState: (newState: T) => void) => DatasourcePublicAPI;
+  // publicAPI: {
+  //   [key in keyof DatasourcePublicAPI]: (
+  //     state: T,
+  //     setState: (newState: T) => void
+  //   ) => DatasourcePublicAPI[key]
+  // };
+}
+
+/**
+ * This is an API provided to visualizations by the frame, which calls the publicAPI on the datasource
+ */
+export interface DatasourcePublicAPI {
+  getTableSpec: () => TableSpec;
+  getOperationForColumnId: () => Operation;
+
+  // Called by dimension
+  getDimensionPanelComponent: (
+    props: DatasourceDimensionPanelProps
+  ) => (domElement: Element, operations: Operation[]) => void;
+
+  removeColumnInTableSpec: (columnId: string) => TableSpec;
+  moveColumnTo: (columnId: string, targetIndex: number) => void;
+  duplicateColumn: (columnId: string) => TableSpec;
+}
+
+export interface DatasourceDataPanelProps {
+  domElement: Element;
+}
+
+// The only way a visualization has to restrict the query building
+export interface DatasourceDimensionPanelProps {
+  // If no columnId is passed, it will render as empty
+  columnId?: string;
+
+  // Visualizations can restrict operations based on their own rules
+  filterOperations: (operation: Operation) => boolean;
+
+  // Visualizations can hint at the role this dimension would play, which
+  // affects the ordering of the query
+  suggestedPriority?: DimensionPriority;
+}
+
+export interface DatasourceValidationError {
+  type: 'error';
+}
+
+export interface DatasourceValidationValid {
+  type: 'valid';
+}
+
+export type DataType = 'string' | 'number' | 'date' | 'boolean';
+
+// An operation does not represent a column in the datatable
```

Comment:
I'd add what it *does* represent to the comment, rather than just a negation. (e.g. // An operation represents a discrete query operation (such as 'count' or 'sum' or 'terms')